pymbar/mbar_solvers.py

```diff
@@ -145,16 +268,27 @@ def mbar_objective_and_gradient(u_kn, N_k, f_k):
     The gradient is equation C6 in the JCP MBAR paper; the objective
     function is its integral.
     """
-    u_kn, N_k, f_k = validate_inputs(u_kn, N_k, f_k)
 
-    log_denominator_n = logsumexp(f_k - u_kn.T, b=N_k, axis=1)
-    log_numerator_k = logsumexp(-log_denominator_n - u_kn, axis=1)
-    grad = -1 * N_k * (1.0 - np.exp(f_k + log_numerator_k))
+    return jit_mbar_objective_and_gradient(u_kn, N_k, f_k)
 
-    obj = math.fsum(log_denominator_n) - N_k.dot(f_k)
+def jax_mbar_hessian(u_kn, N_k, f_k):
+    """JAX version of mbar_hessian.
+    For parameters, see mbar_hessian
+    N_k must be float (should be cast at a higher level)
 
-    return obj, grad
+    """
+    
+    log_denominator_n = logsumexp(f_k - u_kn.T, b=N_k, axis=1)
+    logW = f_k - u_kn.T - log_denominator_n[:, newaxis]
+    W = exp(logW)
 
+    H = dot(W.T,W)
+    H *= N_k
+    H *= N_k[:, newaxis]
+    H -= diag(W.sum(0) * N_k)
+    return -1.0 * H
+
+jit_mbar_hessian = jax.jit(jax_mbar_hessian)
```

Comment:
Generally, best practice is to delay `jit` as late as possible.

Reply:
If you're using `jax.scipy.optimize` the entire optimization routine itself can probably be `jitted`. Generally you want to`jit` only once, and only at the outer most level.

Reply:
I guess the issue is that different branches need to be jitted at different levels.  

I will try to at least jit around the loop in adaptive, or jit the entire function, though!